### PR TITLE
Merkle proof api

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
Added two new endpoints: one for computing a Merkle proof, and another for verifying a Merkle proof.

### CODE REVIEW

File: api/groups/baseProofGroup.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		{Path: "/verify", Handler: pg.verifyProof, Method: http.MethodPost},
+	}
+	pg.baseGroup.endpoints = baseRoutesHandlers
+
+	return pg, nil
+}
+
+func (pg *proofGroup) getProof(c *gin.Context) {
+	rootHash := c.Param("roothash")
+	address := c.Param("address")
```
Comment (bogdan-rosianu) [inline api/groups/baseProofGroup.go:40]: check empty roothash or address?

File: api/groups/baseProofGroup.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+	proof, err := pg.facade.GetProof([]byte(rootHash), []byte(address))
+	if err != nil {
+		shared.RespondWith(c, http.StatusInternalServerError, nil, err.Error(), data.ReturnCodeInternalError)
+		return
+	}
+
+	shared.RespondWith(c, http.StatusOK, gin.H{"proof": proof}, "", data.ReturnCodeSuccess)
```
Comment (bogdan-rosianu) [inline api/groups/baseProofGroup.go:48]: hex encoded?

File: data/proof.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package data
+
+// GetProofResponse represents the response for the GetProof endpoint
+type GetProofResponse struct {
+	Data  [][]byte `json:"data"`
```
Comment (bogdan-rosianu) [inline data/proof.go:5]: i would encode to hex

File: data/proof.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	Data  bool   `json:"data"`
+	Error string `json:"error"`
+	Code  string `json:"code"`
+}
+
+// VerifyProofRequest represents the parameters needed to verify a Merkle proof
+type VerifyProofRequest struct {
+	RootHash []byte   `json:"roothash"`
+	Address  []byte   `json:"address"`
+	Proof    [][]byte `json:"proof"`
```
Comment (bogdan-rosianu) [inline data/proof.go:21]: hex

File: process/proofProcessor.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	return &ProofProcessor{
+		proc:            proc,
+		pubKeyConverter: pubKeyConverter,
+	}, nil
+}
+
+// GetProof sends the request to the right observer and then replies with the returned answer
+func (pp *ProofProcessor) GetProof(rootHash []byte, address []byte) ([][]byte, error) {
```
Comment (bogdan-rosianu) [inline process/proofProcessor.go:32]: if you don't need to change in any way the response from the node, then you could use GenericApiResponse and this way you make sure the response is consistent with the one from the node

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: api/groups/baseProofGroup.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply api/groups/baseProofGroup.go:40]: Done.

File: api/groups/baseProofGroup.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply api/groups/baseProofGroup.go:48]: Done.

File: data/proof.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply data/proof.go:5]: Removed.

File: data/proof.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply data/proof.go:21]: Done.

File: process/proofProcessor.go (hunk 5 above)
Comment (BeniaminDrasovean) [reply process/proofProcessor.go:32]: Done.

File: api/groups/baseProofGroup.go [hunk 6]
```diff
@@ -37,15 +38,40 @@ func NewProofGroup(facadeHandler data.FacadeHandler) (*proofGroup, error) {
...
+		return
+	}
+
+	address := c.Param("address")
+	if address == "" {
+		shared.RespondWith(c, http.StatusBadRequest, nil, errors.ErrEmptyAddress.Error(), data.ReturnCodeRequestError)
+		return
+	}
+
+	getProofResp, err := pg.facade.GetProof([]byte(rootHash), []byte(address))
```
Comment (bogdan-rosianu) [inline api/groups/baseProofGroup.go:52]: no need to cast the `address` or the `rootHash` to `[]byte` because you convert them back durring processing

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: api/groups/baseProofGroup.go (hunk 6 above)
Comment (BeniaminDrasovean) [reply api/groups/baseProofGroup.go]: Refactored.

**Review by bogdan-rosianu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Pull request merged**